src/languages/javascript.js

```diff
@@ -31,29 +31,51 @@ export default function(hljs) {
   const XML_TAG = {
     begin: /<[A-Za-z0-9\\._:-]+/,
     end: /\/[A-Za-z0-9\\._:-]+>|\/>/,
+    // to avoid special cases inside isTrulyOpeningTag
+    simpleSelfClosing: /<[A-Za-z0-9\\._:-]+\s*\/>/,
     /**
      * @param {RegExpMatchArray} match
      * @param {CallbackResponse} response
      */
     isTrulyOpeningTag: (match, response) => {
       const afterMatchIndex = match[0].length + match.index;
       const nextChar = match.input[afterMatchIndex];
-      // nested type?
-      // HTML should not include another raw `<` inside a tag
-      // But a type might: `<Array<Array<number>>`, etc.
-      if (nextChar === "<") {
+      if (
+        // HTML should not include another raw `<` inside a tag
+        // nested type?
+        // `<Array<Array<number>>`, etc.
+        nextChar === "<" ||
+        // the , gives away that this is not HTML
+        // `<T, A extends keyof T, V>`
+        nextChar === ",") {
         response.ignoreMatch();
         return;
       }
-      // <something>
-      // This is now either a tag or a type.
+
+      // `<something>`
+      // Quite possibly a tag, lets look for a matching closing tag...
       if (nextChar === ">") {
         // if we cannot find a matching closing tag, then we
         // will ignore it
         if (!hasClosingTag(match, { after: afterMatchIndex })) {
           response.ignoreMatch();
         }
       }
+
+      // `<blah />` (self-closing)
+      // handled by simpleSelfClosing rule
+
+      // `<From extends string>`
+      // technically this could be HTML, but it smells like a type
+      let m;
+      const afterMatch = match.input.substr(afterMatchIndex)
+      // NOTE: This is ugh, but added specifically for https://github.com/highlightjs/highlight.js/issues/3276
+      if (m = afterMatch.match(/^\s+extends\s+/)) {
```

Comment:
```suggestion
      if ((m = afterMatch.match(/^\s+extends\s+/))) {
```

This is just personal preference and I don't know if we have more of these in the codebase, but I usually surround assignments in if statements with an extra set of `()` so that it signifies it's not a type for `==`.

Reply:
Ugh, why is From and string there highlighted diff in your snap?

Reply:
Ugh, and if we don't flag it as HTML then we fall into the class "X extends Y" rule, which isn't really correct either... this may need more thought.